netket/utils/history.py

```diff
@@ -302,6 +212,59 @@ def __str__(self):
         return f"MVHistory(keys={self.keys()}, n_iters={len(self.iters)})"
 
 
+@dispatch.annotations()
+def append(self: MVHistory, val: MVHistory, it: object = None):
+    if not set(self.keys()) == set(val.keys()):
+        raise ValueError("cannot concatenate MVHistories with different keys")
+
+    for key in self.keys():
+        self._value_dict[key] = np.concatenate([self[key], val[key]])
+
+    self._value_dict["iters"] = np.concatenate([self.iters, val.iters])
+
+    self._len = len(self) + len(val)
+
+
+@dispatch.annotations()
+def append(self: MVHistory, values: dict, it: object = None):
+    for key, val in values.items():
+        _vals = self._value_dict[key]
+
+        if isinstance(_vals, list):
+            _vals.append(val)
+        elif isinstance(_vals, np.ndarray):
+            new_shape = (len(_vals) + 1,) + _vals.shape[1:]
+            try:
+                _vals.resize(new_shape)
+            except:
+                _vals = np.resize(_vals, new_shape)
+                self._value_dict[key] = _vals
+
+            _vals[-1] = val
+        else:
+            raise TypeError(f"Unknown accumulator type {type(_vals)} for key {key}.")
```

Comment:
Maybe I'm tired and missing something obvious, but why is this not just `np.concatenate` or `np.append`?

Reply:
Append works only for 1d array so I cannot use that.
In principle I could concatenate, but since this is run at every iteration, while not particularly expensive, I wanted to avoid reallocating the array at every step.

`vals.resize(new_shape)` attempts to resize in-place the buffer (if there is enough free space) and reallocates to a double-sized buffer if it fails, so that it then won't need to reallocate for some time.

The try-except is because this trick does not work if you are holding some reference to `vals` when you run it, and it throws. In those cases I use `np.resize` that does the same thing, but guarantees reallocation.